File: src/main/java/com/unihogsoft/scalag/vulkan/VulkanContext.java
package com.unihogsoft.scalag.vulkan;

import com.unihogsoft.scalag.vulkan.command.CommandPool;
import com.unihogsoft.scalag.vulkan.command.Queue;
import com.unihogsoft.scalag.vulkan.command.StandardCommandPool;
import com.unihogsoft.scalag.vulkan.core.DebugCallback;
import com.unihogsoft.scalag.vulkan.core.Device;
import com.unihogsoft.scalag.vulkan.core.Instance;
import com.unihogsoft.scalag.vulkan.memory.Allocator;
import com.unihogsoft.scalag.vulkan.memory.DescriptorPool;

/**
 * @author MarconZet
 * Created 13.04.2020
 */
public class VulkanContext {
    public static final String[] VALIDATION_LAYERS = {"VK_LAYER_KHRONOS_validation"};
    private final boolean enableValidationLayers;

    private Instance instance;
    private DebugCallback debugCallback;
    private Device device;
    private Allocator allocator;
    private Queue computeQueue;
    private DescriptorPool descriptorPool;
    private CommandPool commandPool;

    public VulkanContext() {
        this(false);
    }

    public VulkanContext(boolean enableValidationLayers) {
        this.enableValidationLayers = enableValidationLayers;
        instance = new Instance(enableValidationLayers);
        if(enableValidationLayers)
            debugCallback = new DebugCallback(instance);
        device = new Device(enableValidationLayers, instance);
        computeQueue = new Queue(device.getComputeQueueFamily(), 0, device);
        allocator = new Allocator(instance, device);
        descriptorPool = new DescriptorPool(device);
        commandPool = new StandardCommandPool(device, computeQueue);
    }

    public void close(){
        commandPool.destroy();
        descriptorPool.destroy();
        allocator.destroy();
        computeQueue.destroy();
        device.destroy();
        if(enableValidationLayers)
            debugCallback.destroy();
        instance.destroy();
    }



    public Device getDevice() {
        return device;
    }

    public Allocator getAllocator() {
        return allocator;
    }

    public Queue getComputeQueue() {
        return computeQueue;
    }

    public DescriptorPool getDescriptorPool() {
        return descriptorPool;
    }

    public CommandPool getCommandPool() {
        return commandPool;
    }
}

File: src/main/scala/com/scalag/vulkan/VulkanContext.scala
package com.scalag.vulkan

import com.scalag.vulkan.command.{CommandPool, Queue, StandardCommandPool}
import com.scalag.vulkan.core.{DebugCallback, Device, Instance}
import com.scalag.vulkan.memory.{Allocator, DescriptorPool}

/** @author
  *   MarconZet Created 13.04.2020
  */
object VulkanContext {
  final val ValidationLayer: String = "VK_LAYER_KHRONOS_validation"
}

class VulkanContext(val enableValidationLayers: Boolean = false) {
  private val validationLayers = enableValidationLayers || true

  val instance: Instance = new Instance(validationLayers)
  val debugCallback: DebugCallback = if (validationLayers) new DebugCallback(instance) else null
  val device: Device = new Device(instance)
  val computeQueue: Queue = new Queue(device.computeQueueFamily, 0, device)
  val allocator: Allocator = new Allocator(instance, device)
  val descriptorPool: DescriptorPool = new DescriptorPool(device)
  val commandPool: CommandPool = new StandardCommandPool(device, computeQueue)

  def destroy(): Unit = {
    commandPool.destroy()
    descriptorPool.destroy()
    allocator.destroy()
    computeQueue.destroy()
    device.destroy()
    if (validationLayers)
      debugCallback.destroy()
    instance.destroy()
  }
}
